tests/ComApartmentVariableTests.cpp

```diff
@@ -0,0 +1,427 @@
+    ULONG index{};
+    FAIL_FAST_IF_FAILED(CoWaitForMultipleHandles(COWAIT_DISPATCH_CALLS, INFINITE, static_cast<ULONG>(std::size(raw)), raw, &index));
+}
+
+void RunApartmentVariableTest(void(*test)())
+{
+    test();
+    // Apartment variable rundown is async, wait for the last COM apartment
+    // to rundown before proceeding to the next test.
+    WaitForAllComApartmentsToRundown();
+}
+
+struct mock_platform
+{
+    static unsigned long long GetApartmentId()
+    {
+        APTTYPE type; APTTYPEQUALIFIER qualifer;
+        FAIL_FAST_IF_FAILED(CoGetApartmentType(&type, &qualifer)); // ensure COM is inited
+
+        // Approximate apartment Id
+        if (type == APTTYPE_STA)
+        {
+            FAIL_FAST_IF(GetCurrentThreadId() < APTTYPE_MAINSTA);
+            return GetCurrentThreadId();
+        }
+        else
+        {
+            // APTTYPE_MTA (1), APTTYPE_NA (2), APTTYPE_MAINSTA (3)
+            return type;
+        }
+    }
+
+    static auto RegisterForApartmentShutdown(IApartmentShutdown* observer)
+    {
+        const auto id = GetApartmentId();
+        auto apt_observers = m_observers.find(id);
+        if (apt_observers == m_observers.end())
+        {
+            m_observers.insert({ id, { observer} });
+        }
+        else
+        {
+            apt_observers->second.push_back(observer);
+        }
+        return shutdown_type{ reinterpret_cast<APARTMENT_SHUTDOWN_REGISTRATION_COOKIE>(id) };
+    }
+
+    static void UnRegisterForApartmentShutdown(APARTMENT_SHUTDOWN_REGISTRATION_COOKIE cookie)
+    {
+        auto id = reinterpret_cast<unsigned long long>(cookie);
+        m_observers.erase(id);
+    }
+
+    using shutdown_type = wil::unique_any<APARTMENT_SHUTDOWN_REGISTRATION_COOKIE, decltype(&UnRegisterForApartmentShutdown), UnRegisterForApartmentShutdown>;
+
+    // This is needed to simulate the platform for unit testing.
+    static auto CoInitializeEx(DWORD coinitFlags = 0 /*COINIT_MULTITHREADED*/)
+    {
+        return wil::scope_exit([aptId = GetCurrentThreadId(), init = wil::CoInitializeEx(coinitFlags)]()
+        {
+            const auto id = GetApartmentId();
+            auto apt_observers = m_observers.find(id);
+            if (apt_observers != m_observers.end())
+            {
+                const auto& observers = apt_observers->second;
+                for (auto& observer : observers)
+                {
+                    observer->OnUninitialize(id);
+                }
+                m_observers.erase(apt_observers);
+            }
+        });
+    }
+
+    // Enable the test hook to force losing the race
+    inline static constexpr unsigned long AsyncRundownDelayForTestingRaces = 1; // enable test hook
+    inline static std::unordered_map<unsigned long long, std::vector<wil::com_ptr<IApartmentShutdown>>> m_observers;
+};
+
+auto fn() { return 42; };
+auto fn2() { return 43; };
+
+wil::apartment_variable<int, mock_platform> g_v1;
+wil::apartment_variable<int> g_v2;

```

Comment:
I'm a bit confused by these. g_v2 appears to be completely unreferenced. g_v1 is referenced only from TestApartmentVariableAllMethods, and only minimally there (and uses mock_platform even in the UseRealPlatformRunAllTests case). Are these just meant to smoke-test the global variable-specific stuff, or are they potentially incomplete (or am I just missing something)?

Reply:
smoke test of global variables.